### PR TITLE
chore(ext/webgpu): include GPUCanvasContext in snapshot

### PR DESCRIPTION
**Pull request by littledivy**
Part 1 of #21713 

Changes:

- Remove `.present()` and add a `presentGPUCanvasContext` (not exposed yet to users)
- Move lazy load logic to `00_init.js`. This can be used to use webgpu on-demand from future code (OffScreenCanvas)

### CODE REVIEW

File: ext/webgpu/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ description = "WebGPU implementation for Deno"
 path = "lib.rs"
 
 [features]
+default = ["surface"]
```
Comment (crowlKats) [inline ext/webgpu/Cargo.toml:17]: remove the surface feature

**Comment by crowlKats** [comment]
why is `00_init.js` necessary?

**Comment by littledivy** [comment]
> why is 00_init.js necessary?

So that webgpu can be loaded from any part of the codebase. In this case, `02_surface.js` needs to use webgpu internals.

**Comment by crowlKats** [comment]
a lazy module should be able to import another lazy module without sucha. workaorund, just another import. if this currently does not work, thats an issue

**Comment by littledivy** [comment]
Yes `op_lazy_load_esm` works from anywhere. 

The problem is that we need to share the `webgpu` variable to avoid double loading:
```js
function loadWebGPU() {
  if (!webgpu) {
    webgpu = ops.op_lazy_load_esm("ext:deno_webgpu/01_webgpu.js");
  }
}
```

**Comment by crowlKats** [comment]
Ah I see only now that you are not lazy loading surface. it definitively should be lazy loaded.

**Comment by littledivy** [comment]
> it definitively should be lazy loaded.

Why is it worth lazy loading a 150LoC file if we already have most of the snapshot-heavy code in 01_webgpu.js?

if your concern is `OffScreenCanvas`, how is "not-lazy loading" 02_surface.js bad?

From what I understand here, this will only increase overhead.

**Comment by crowlKats** [comment]
a) we need to start lazy loading existing modules that are not in a hot path, and as such this applies to new ones as well.
b) this file will grow, and given that are doing some "weird" lazy loading things here, we should do it properly from the beginning.

> Yes `op_lazy_load_esm` works from anywhere.

I meant a static import: a lazy module static importing another lazy module should work without issue. if it doesnt return the same object as the lazy load op, thats also an issue that needs to be addressed

**Comment by littledivy** [comment]
> a) we need to start lazy loading existing modules that are not in a hot path, and as such this applies to new ones as well.

disagree. lazy loading is _only_ beneficial for large singular source files where the snapshot takes longer to deserialize. Startup time matters for desktop apps equally, I would not consider this a non-hot path and call it a day.

"this file will grow" is not a good reason to add overhead before it starts growing. If we really want to lazy load surface.js, we should do it when its worth it with proper benchmarks - not now.

For now, I would say we do not lazy-load surface.js unless its actually a problem.

**Comment by littledivy** [comment]
Ok so I benchmarked startup time with lazy surface:

```js
// test.js
globalThis.GPUCanvasContext
```

```
divy@mini ~/g/deno (byow_part_1) [1]> hyperfine './deno2 run -A --unstable test.js' 'target/release/deno run -A --unstable test.js'
Benchmark 1: ./deno2 run -A --unstable test.js
  Time (mean ± σ):      16.0 ms ±   2.3 ms    [User: 11.1 ms, System: 4.7 ms]
  Range (min … max):    14.9 ms …  39.3 ms    127 runs

Benchmark 2: target/release/deno run -A --unstable test.js
  Time (mean ± σ):      18.5 ms ±   0.6 ms    [User: 13.6 ms, System: 4.6 ms]
  Range (min … max):    17.9 ms …  22.3 ms    115 runs

Summary
  'deno run -A --unstable test.js' ran
    1.16 ± 0.17 times faster than 'target/release/deno run -A --unstable test.js'
```

That's over 1ms overhead with an extra lazy load happening.

@crowlKats If you still think its better to lazy load 02_surface.js, I'll push the changes.

**Comment by bartlomieju** [comment]
Does surface require for WebGPU to be loaded upfront?

**Comment by littledivy** [comment]
> Does surface require for WebGPU to be loaded upfront?

Yes, and 02_surface.js will load 01_webgpu.js unless its already loaded before.

**Comment by bartlomieju** [comment]
> > Does surface require for WebGPU to be loaded upfront?
> 
> Yes, and 02_surface.js will load 01_webgpu.js unless its already loaded before.

So that means that `01_webgpu.js` will be loaded unconditionally on each startup. I'm surprised that this is faster to be included in the snapshot that lazy loaded when needed.

**Comment by littledivy** [comment]
> So that means that 01_webgpu.js will be loaded unconditionally on each startup. 

Nope, 02_surface.js only loads 01_webgpu.js when it is needed, not unconditionally.

**Comment by crowlKats** [comment]
@littledivy what is the code you did for the lazy scenario?

**Comment by littledivy** [comment]
Just this:
```
globalThis.GPUCanvasContext
```

It's above the benchmark results. https://github.com/denoland/deno/pull/21773#issuecomment-1875187619

**Comment by crowlKats** [comment]
ah so you didnt actually change any of the JS code of 02_surface? Then not surprising given that you are lazy loading surface, which then lazy loads webgpu

**Comment by crowlKats** [comment]
ah so you didnt change any of the js in 02_surface?

**Comment by littledivy** [comment]
Ah no I did, I thought you meant the benchmark code actually used.

Lemme push that to another branch

**Comment by littledivy** [comment]
@crowlKats Patch for the lazy surface scenario: https://github.com/denoland/deno/commit/6425f95c90cfb72733266e24d94574957ff5a9c6 (It is static importing webgpu)

**Comment by crowlKats** [comment]
I would say the benchmark is not a real scenario: the test code is just accessing the class (which no one will actually do), which in the lazy scenario will load webgpu, whereas in the non-lazy it doesnt, which obviously will make a big difference. A real scenario of this will be actually using it

**Comment by littledivy** [comment]
It's emulating a real world case. This is what a real-world would do, it has same effect of lazy loading 2 sources:

```js
const adapter = await navigator.gpu.requestAdapter(); // Causes 01_webgpu.js to lazy load
const ctx = Deno.createWindowSurface(...) //  Causes 02_surface.js to lazy load.
```

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**